drivers/flash/flash_ifx_cat1.c

```diff
@@ -0,0 +1,193 @@
+/*
+ * Copyright 2023 Cypress Semiconductor Corporation (an Infineon company) or
+ * an affiliate of Cypress Semiconductor Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT	  infineon_cat1_flash_controller
+#define SOC_NV_FLASH_NODE DT_PARENT(DT_INST(0, fixed_partitions))
+
+#define PAGE_LEN   DT_PROP(SOC_NV_FLASH_NODE, erase_block_size)
+#define PAGE_LEN_P 9
+
+#include <zephyr/kernel.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/flash.h>
+#include <zephyr/logging/log.h>
+
+#include "cyhal_flash.h"
+
+LOG_MODULE_REGISTER(flash_infineon_cat1, CONFIG_FLASH_LOG_LEVEL);
+
+/* Device config structure */
+struct ifx_cat1_flash_config {
+	uint32_t base_addr;
+};
+
+/* Data structure */
+struct ifx_cat1_flash_data {
+	cyhal_flash_t flash_obj;
+	struct k_sem sem;
+};
+
+static struct flash_parameters ifx_cat1_flash_parameters = {
+	.write_block_size = DT_PROP(SOC_NV_FLASH_NODE, write_block_size),
+	.erase_value = 0x00,
+};
+
+static inline void flash_ifx_sem_take(const struct device *dev)
+{
+	struct ifx_cat1_flash_data *data = dev->data;
+
+	k_sem_take(&data->sem, K_FOREVER);
+}
+
+static inline void flash_ifx_sem_give(const struct device *dev)
+{
+	struct ifx_cat1_flash_data *data = dev->data;
+
+	k_sem_give(&data->sem);
+}
+
+static int ifx_cat1_flash_read(const struct device *dev, off_t address, void *buffer, size_t length)
+{
+	struct ifx_cat1_flash_data *data = dev->data;
+	const struct ifx_cat1_flash_config *config = dev->config;
+
+	flash_ifx_sem_take(dev);
+
+	cy_rslt_t rslt = cyhal_flash_read(&data->flash_obj, address + config->base_addr,
+					  (uint8_t *)buffer, length);
+	if (rslt != CY_RSLT_SUCCESS) {
+		flash_ifx_sem_give(dev);
+		return -EIO;
+	}
+	flash_ifx_sem_give(dev);
+	return 0;
+}
+
+static int _ifx_cat1_flash_write(cyhal_flash_t *flash_obj, uint32_t write_addr, uint32_t *buf,
+				 size_t length)
+{
+	uint8_t local_buffer[PAGE_LEN];
+
+	/* Start address of page */
+	uint32_t page_start_addr = (write_addr >> PAGE_LEN_P) << PAGE_LEN_P;
```

Comment:
I do not think that write address should be corrected when it does not align with page size. User may wrongly set start address but that should be rater reported as error, otherwise this just removes some random stuff - something we may suspect user for, but driver should not do.
And the check is in wrong place, because, I assume, that incorrect write address may be already passed to `ifx_cat1_flash_write`

Reply:
Per the below documentation for the flash_write, its not necessary that the source address is aligned to the page size. Also, we read the entire page, modify the bytes to be updated and then write the entire page back, so we are not affecting the existing data..

Reply:
> Per the below documentation for the flash_write, its not necessary that the source address is aligned to the page size. Also, we read the entire page, modify the bytes to be updated and then write the entire page back, so we are not affecting the existing data..

Yes that is correct, this means that offset of source may be anything and driver should be able to deal with this. For example if you have some DMA support that only reads from uint32 aligned addresses and user gives you `offset % 4 != 0` it is drivers job to copy data to intermediate buffer and deal with this.
But when user requests write flash write offset to be something driver may not deal with, driver should report error, not correct that.

Reply:
This is from your code
```
static struct flash_parameters ifx_cat1_flash_parameters = {
	.write_block_size = DT_PROP(SOC_NV_FLASH_NODE, write_block_size),
	.erase_value = 0x00,
};
```
so as far as I can tell your device requires write alignment defined by `write-block-size property`, not `PAGE_LEN`, which seems to be `erase-page-size` property.
The `write-block-size` property, as defined here dts/arm/infineon/psoc6/mpns/CY8C624ABZI_S2D44.dtsi, is 4 for the device. Which means that user, while calling the `flash_write` should provide 4 bytes of data with 4 byte alignment in target offset, otherwise error. I assume, by lack of length parameter for `cyhal_flash_write` that there is error in dtsi and write alignment should be equal to erase page size (512).
Your driver call behaves differently than expected from the documentation, because it corrects destination offset, something it should not do.
Does your device allow rewrites of already written bytes?
You provide erase function so I guess that erase is needed between writes?

Reply:
I misunderstood the documentation for the flash_write API, I have now updated the write-block-size to 512 and removed the logic to correct the offset.

When testing this using the nvs application, I have realized that there is the following check in the nvs sub system that restricts write-block-size(NVS_BLOCK_SIZE) to 32. Would you know the reason for this?

```
    /* check that the write block size is supported */
    if (write_block_size > NVS_BLOCK_SIZE || write_block_size == 0) {
        LOG_ERR("Unsupported write block size");
        return -EINVAL;
    } 
```

Reply:
Because NVS record, or whatever you call it, is of certain size. If your write alignment is higher than that block, this means you would waste flash using NVS, for example if NVS block is 32 bytes and you write block is 512 bytes, you would have to use 512 size block entirely to store the 32 bytes.

Reply:
so should we increase the NVS_BLOCK_SIZE to 512 when using our board?